File: backend/src/controllers/product.ts
import { Request, Response } from 'express';
import prisma from '../config/db';
import cloudinary from 'cloudinary';

cloudinary.v2.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
});

// Helper function to ensure single string
const toString = (value: any): string => {
  return Array.isArray(value) ? value[0] : (value as string);
};

// GET /api/products
export const getProducts = async (req: Request, res: Response): Promise<void> => {
  try {
    const { category, search } = req.query;

    const products = await prisma.product.findMany({
      where: {
        shopId: req.shopId as string,
        isActive: true,
        ...(category ? { category: toString(category) } : {}),
        ...(search
          ? {
            OR: [
              { name: { contains: toString(search) } },
              { sku: { contains: toString(search) } },
              { barcode: { contains: toString(search) } },
            ],
          }
          : {}),
      },
      orderBy: { name: 'asc' },
    });

    res.json(products);
  } catch (error) {
    console.error('Get products error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// GET /api/products/:id
export const getProduct = async (req: Request, res: Response): Promise<void> => {
  try {
    const productId = toString(req.params.id);

    const product = await prisma.product.findFirst({
      where: {
        id: productId,
        shopId: req.shopId as string
      },
    });

    if (!product) {
      res.status(404).json({ error: 'Product not found' });
      return;
    }

    res.json(product);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
};

// POST /api/products
export const createProduct = async (req: Request, res: Response): Promise<void> => {
  try {
    const { name, description, price, stock, sku, category, imageUrl } = req.body;

    if (!name || price === undefined) {
      res.status(400).json({ error: 'name and price are required' });
      return;
    }

    const shopId = req.shopId as string;

    // Enforce product limit from subscription plan
    const subscription = await prisma.shopSubscription.findFirst({
      where: { shopId, status: 'active' },
      include: { plan: { select: { name: true, productsLimit: true } } },
      orderBy: { createdAt: 'desc' },
    });

    if (subscription) {
      const limit = subscription.plan.productsLimit;
      const existingCount = await prisma.product.count({
        where: { shopId, isActive: true },
      });

      if (existingCount >= limit) {
        res.status(403).json({
          error: `Product limit reached. Your ${subscription.plan.name} plan allows ${limit} products. You have ${existingCount}/${limit}.`,
        });
        return;
      }
    }

    const product = await prisma.product.create({
      data: {
        shopId,
        name,
        description: description || '',
        price: parseFloat(price),
        stock: parseInt(stock) || 0,
        sku: sku || '',
        category: category || 'General',
        imageUrl: imageUrl || '',
        barcode: req.body.barcode || null,
        lowStockThreshold: parseInt(req.body.lowStockThreshold) || 5,
        batchNumber: req.body.batchNumber || null,
        expiryDate: req.body.expiryDate ? new Date(req.body.expiryDate) : null,
        manufacturer: req.body.manufacturer || null,
        composition: req.body.composition || null,
        dosageForm: req.body.dosageForm || null,
        packing: req.body.packing || null,
        isControlled: req.body.isControlled ?? false,
        supplierId: req.body.supplierId || null,
      },
    });

    res.status(201).json(product);
  } catch (error) {
    console.error('Create product error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// PUT /api/products/:id
export const updateProduct = async (req: Request, res: Response): Promise<void> => {
  try {
    const productId = toString(req.params.id);
    const shopId = req.shopId as string;

    const existing = await prisma.product.findFirst({
      where: { id: productId, shopId: shopId },
    });

    if (!existing) {
      res.status(404).json({ error: 'Product not found' });
      return;
    }

    const { name, description, price, stock, sku, category, imageUrl, isActive, barcode, lowStockThreshold } = req.body;

    const product = await prisma.product.update({
      where: { id: productId },
      data: {
        ...(name !== undefined && { name }),
        ...(description !== undefined && { description }),
        ...(price !== undefined && { price: parseFloat(price) }),
        ...(stock !== undefined && { stock: parseInt(stock) }),
        ...(sku !== undefined && { sku }),
        ...(category !== undefined && { category }),
        ...(imageUrl !== undefined && { imageUrl }),
        ...(isActive !== undefined && { isActive }),
        ...(barcode !== undefined && { barcode: barcode || null }),
        ...(lowStockThreshold !== undefined && { lowStockThreshold: parseInt(lowStockThreshold) }),
        ...(req.body.batchNumber !== undefined && { batchNumber: req.body.batchNumber }),
        ...(req.body.expiryDate !== undefined && { expiryDate: req.body.expiryDate ? new Date(req.body.expiryDate) : null }),
        ...(req.body.manufacturer !== undefined && { manufacturer: req.body.manufacturer }),
        ...(req.body.composition !== undefined && { composition: req.body.composition }),
        ...(req.body.dosageForm !== undefined && { dosageForm: req.body.dosageForm }),
        ...(req.body.packing !== undefined && { packing: req.body.packing }),
        ...(req.body.isControlled !== undefined && { isControlled: req.body.isControlled }),
        ...(req.body.supplierId !== undefined && { supplierId: req.body.supplierId }),
      },
    });

    res.json(product);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
};

// DELETE /api/products/:id
export const deleteProduct = async (req: Request, res: Response): Promise<void> => {
  try {
    const productId = toString(req.params.id);
    const shopId = req.shopId as string;

    const existing = await prisma.product.findFirst({
      where: { id: productId, shopId: shopId },
    });

    if (!existing) {
      res.status(404).json({ error: 'Product not found' });
      return;
    }

    await prisma.product.update({
      where: { id: productId },
      data: { isActive: false },
    });

    res.json({ message: 'Product deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
};

// GET /api/products/low-stock — products where stock <= lowStockThreshold
export const getLowStockProducts = async (req: Request, res: Response): Promise<void> => {
  try {
    const shopId = req.shopId as string;
    const products = await prisma.product.findMany({
      where: { shopId, isActive: true },
      select: { id: true, name: true, stock: true, lowStockThreshold: true, sku: true, price: true, imageUrl: true, shopId: true, createdAt: true },
    });
    const lowStock = products.filter((p) => p.stock <= p.lowStockThreshold);
    res.json({ count: lowStock.length, products: lowStock });
  } catch (error) {
    console.error('Get low stock error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// GET /api/products/categories
export const getCategories = async (req: Request, res: Response): Promise<void> => {
  try {
    const products = await prisma.product.findMany({
      where: { shopId: req.shopId as string, isActive: true },
      select: { category: true },
      distinct: ['category'],
    });

    res.json(products.map((p) => p.category));
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
};

// GET /api/products/barcode/:barcode
export const getProductByBarcode = async (req: Request, res: Response): Promise<void> => {
  try {
    const barcode = toString(req.params.barcode);
    console.log('BARCODE LOOKUP - received:', JSON.stringify(barcode));

    if (!barcode) {
      res.status(400).json({ error: 'Barcode is required' });
      return;
    }

    const product = await prisma.product.findFirst({
      where: {
        barcode: barcode,
        shopId: req.shopId as string,
        isActive: true,
      },
    });

    if (!product) {
      res.status(404).json({ error: 'Product not found' });
      return;
    }

    res.json(product);
  } catch (error) {
    console.error('Get product by barcode error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// POST /api/products/upload-image
export const uploadImage = async (req: Request, res: Response): Promise<void> => {
  try {
    if (!req.file) {
      res.status(400).json({ error: 'No image file provided' });
      return;
    }

    const { CLOUDINARY_CLOUD_NAME, CLOUDINARY_API_KEY, CLOUDINARY_API_SECRET } = process.env;
    if (!CLOUDINARY_CLOUD_NAME || !CLOUDINARY_API_KEY || !CLOUDINARY_API_SECRET) {
      res.status(500).json({ error: 'Cloudinary not configured on server' });
      return;
    }

    const result = await new Promise<any>((resolve, reject) => {
      const stream = cloudinary.v2.uploader.upload_stream(
        { folder: 'bluexsofts/products', resource_type: 'image' },
        (error, result) => {
          if (error) reject(error);
          else resolve(result);
        },
      );
      stream.end(req.file!.buffer);
    });

    res.json({ imageUrl: result.secure_url });
  } catch (error) {
    console.error('Upload image error:', error);
    res.status(500).json({ error: 'Failed to upload image' });
  }
};

// ── Category-specific product endpoints ────────────────────────────

// Pharmacy: products expiring within N days
export const getExpiringProducts = async (req: Request, res: Response): Promise<void> => {
  try {
    const days = parseInt(toString(req.query.days)) || 30;
    const shopId = req.shopId as string;

    const threshold = new Date();
    threshold.setDate(threshold.getDate() + days);

    const products = await prisma.product.findMany({
      where: {
        shopId,
        isActive: true,
        expiryDate: { not: null, lte: threshold },
      },
      orderBy: { expiryDate: 'asc' },
    });

    res.json(products);
  } catch (error) {
    console.error('getExpiringProducts error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// Pharmacy: count of expiring products
export const getExpiringCount = async (req: Request, res: Response): Promise<void> => {
  try {
    const days = parseInt(toString(req.query.days)) || 30;
    const shopId = req.shopId as string;

    const threshold = new Date();
    threshold.setDate(threshold.getDate() + days);

    const count = await prisma.product.count({
      where: {
        shopId,
        isActive: true,
        expiryDate: { not: null, lte: threshold },
      },
    });

    res.json({ count });
  } catch (error) {
    console.error('getExpiringCount error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// Electronics: warranty tracking
export const getWarrantyProducts = async (req: Request, res: Response): Promise<void> => {
  try {
    const shopId = req.shopId as string;

    const products = await prisma.product.findMany({
      where: {
        shopId,
        isActive: true,
        warrantyMonths: { not: null },
      },
      orderBy: { name: 'asc' },
    });

    // Compute warranty status for each product
    const now = new Date();
    const withWarranty = products.map(p => {
      let warrantyEnd: Date | null = null;
      let isExpired = false;
      let monthsLeft = 0;
      if (p.warrantyMonths && p.createdAt) {
        warrantyEnd = new Date(p.createdAt);
        warrantyEnd.setMonth(warrantyEnd.getMonth() + p.warrantyMonths);
        isExpired = now > warrantyEnd;
        monthsLeft = !isExpired
          ? Math.round((warrantyEnd.getTime() - now.getTime()) / (30 * 24 * 60 * 60 * 1000))
          : 0;
      }
      return { ...p, warrantyEnd: warrantyEnd?.toISOString(), isExpired, monthsLeft };
    });

    res.json(withWarranty);
  } catch (error) {
    console.error('getWarrantyProducts error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// Grocery: unit-based low stock alerts
export const getUnitLowStock = async (req: Request, res: Response): Promise<void> => {
  try {
    const shopId = req.shopId as string;

    const products = await prisma.product.findMany({
      where: {
        shopId,
        isActive: true,
        unitType: { not: null },
      },
      orderBy: { name: 'asc' },
    });

    const lowStock = products.filter(p => p.stock <= (p.lowStockThreshold || 5));

    res.json(lowStock);
  } catch (error) {
    console.error('getUnitLowStock error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// Restaurant: menu items view
export const getMenuItems = async (req: Request, res: Response): Promise<void> => {
  try {
    const shopId = req.shopId as string;

    const items = await prisma.product.findMany({
      where: {
        shopId,
        isActive: true,
        isMenuItem: true,
      },
      orderBy: { name: 'asc' },
    });

    res.json(items);
  } catch (error) {
    console.error('getMenuItems error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

// Clothing: season-based inventory
export const getSeasonalProducts = async (req: Request, res: Response): Promise<void> => {
  try {
    const shopId = req.shopId as string;
    const season = toString(req.query.season) || '';

    const where: Record<string, unknown> = {
      shopId,
      isActive: true,
    };
    if (season) where.season = season;

    const products = await prisma.product.findMany({
      where: where as any,
      orderBy: { name: 'asc' },
    });

    res.json(products);
  } catch (error) {
    console.error('getSeasonalProducts error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

File: bluexsofts_pos/lib/models/product.dart
class Product {
  final String id;
  final String shopId;
  final String name;
  final String description;
  final double price;
  int stock;
  final String sku;
  final String category;
  final String imageUrl;
  final String? barcode;
  int lowStockThreshold;
  final bool isActive;
  final DateTime createdAt;

  // Grocery
  final String? unitType;
  final double? unitValue;

  // Electronics
  final String? imei;
  final int? warrantyMonths;
  final String? brand;
  final String? model;

  // Restaurant
  final bool isMenuItem;
  final String? recipe;

  // Pharmacy
  final String? batchNumber;
  final DateTime? expiryDate;
  final String? manufacturer;
  final String? composition;
  final String? dosageForm;
  final String? packing;
  final bool isControlled;
  final bool isPrescriptionOnly;

  // Clothing
  final String? size;
  final String? color;
  final String? season;

  // Supplier
  final String? supplierId;

  Product({
    required this.id,
    required this.shopId,
    required this.name,
    this.description = '',
    required this.price,
    this.stock = 0,
    this.sku = '',
    this.category = 'General',
    this.imageUrl = '',
    this.barcode,
    this.lowStockThreshold = 5,
    this.isActive = true,
    required this.createdAt,
    this.unitType,
    this.unitValue,
    this.imei,
    this.warrantyMonths,
    this.brand,
    this.model,
    this.isMenuItem = false,
    this.recipe,
    this.batchNumber,
    this.expiryDate,
    this.manufacturer,
    this.composition,
    this.dosageForm,
    this.packing,
    this.isControlled = false,
    this.isPrescriptionOnly = false,
    this.size,
    this.color,
    this.season,
    this.supplierId,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      shopId: json['shopId'] as String,
      name: json['name'] as String,
      description: json['description'] as String? ?? '',
      price: (json['price'] as num).toDouble(),
      stock: json['stock'] as int? ?? 0,
      sku: json['sku'] as String? ?? '',
      category: json['category'] as String? ?? 'General',
      imageUrl: json['imageUrl'] as String? ?? '',
      barcode: json['barcode'] as String?,
      lowStockThreshold: json['lowStockThreshold'] as int? ?? 5,
      isActive: json['isActive'] as bool? ?? true,
      createdAt: DateTime.parse(json['createdAt'] as String),
      unitType: json['unitType'] as String?,
      unitValue: (json['unitValue'] as num?)?.toDouble(),
      imei: json['imei'] as String?,
      warrantyMonths: json['warrantyMonths'] as int?,
      brand: json['brand'] as String?,
      model: json['model'] as String?,
      isMenuItem: json['isMenuItem'] as bool? ?? false,
      recipe: json['recipe'] as String?,
      batchNumber: json['batchNumber'] as String?,
      expiryDate: json['expiryDate'] != null ? DateTime.parse(json['expiryDate'] as String) : null,
      manufacturer: json['manufacturer'] as String?,
      composition: json['composition'] as String?,
      dosageForm: json['dosageForm'] as String?,
      packing: json['packing'] as String?,
      isControlled: json['isControlled'] as bool? ?? false,
      isPrescriptionOnly: json['isPrescriptionOnly'] as bool? ?? false,
      size: json['size'] as String?,
      color: json['color'] as String?,
      season: json['season'] as String?,
      supplierId: json['supplierId'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'shopId': shopId,
    'name': name,
    'description': description,
    'price': price,
    'stock': stock,
    'sku': sku,
    'category': category,
    'imageUrl': imageUrl,
    'barcode': barcode,
    'lowStockThreshold': lowStockThreshold,
    'isActive': isActive,
    'createdAt': createdAt.toIso8601String(),
    if (unitType != null) 'unitType': unitType,
    if (unitValue != null) 'unitValue': unitValue,
    if (imei != null) 'imei': imei,
    if (warrantyMonths != null) 'warrantyMonths': warrantyMonths,
    if (brand != null) 'brand': brand,
    if (model != null) 'model': model,
    'isMenuItem': isMenuItem,
    if (recipe != null) 'recipe': recipe,
    if (batchNumber != null) 'batchNumber': batchNumber,
    if (expiryDate != null) 'expiryDate': expiryDate!.toIso8601String(),
    if (manufacturer != null) 'manufacturer': manufacturer,
    if (composition != null) 'composition': composition,
    if (dosageForm != null) 'dosageForm': dosageForm,
    if (packing != null) 'packing': packing,
    'isControlled': isControlled,
    'isPrescriptionOnly': isPrescriptionOnly,
    if (size != null) 'size': size,
    if (color != null) 'color': color,
    if (season != null) 'season': season,
    if (supplierId != null) 'supplierId': supplierId,
  };

  Product copyWith({
    String? name,
    String? description,
    double? price,
    int? stock,
    String? sku,
    String? category,
    String? imageUrl,
    String? barcode,
    int? lowStockThreshold,
    bool? isActive,
    String? unitType,
    double? unitValue,
    String? imei,
    int? warrantyMonths,
    String? brand,
    String? model,
    bool? isMenuItem,
    String? recipe,
    String? batchNumber,
    DateTime? expiryDate,
    String? manufacturer,
    String? composition,
    String? dosageForm,
    String? packing,
    bool? isControlled,
    bool? isPrescriptionOnly,
    String? size,
    String? color,
    String? season,
    String? supplierId,
  }) {
    return Product(
      id: id,
      shopId: shopId,
      name: name ?? this.name,
      description: description ?? this.description,
      price: price ?? this.price,
      stock: stock ?? this.stock,
      sku: sku ?? this.sku,
      category: category ?? this.category,
      imageUrl: imageUrl ?? this.imageUrl,
      barcode: barcode ?? this.barcode,
      lowStockThreshold: lowStockThreshold ?? this.lowStockThreshold,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt,
      unitType: unitType ?? this.unitType,
      unitValue: unitValue ?? this.unitValue,
      imei: imei ?? this.imei,
      warrantyMonths: warrantyMonths ?? this.warrantyMonths,
      brand: brand ?? this.brand,
      model: model ?? this.model,
      isMenuItem: isMenuItem ?? this.isMenuItem,
      recipe: recipe ?? this.recipe,
      batchNumber: batchNumber ?? this.batchNumber,
      expiryDate: expiryDate ?? this.expiryDate,
      manufacturer: manufacturer ?? this.manufacturer,
      composition: composition ?? this.composition,
      dosageForm: dosageForm ?? this.dosageForm,
      packing: packing ?? this.packing,
      isControlled: isControlled ?? this.isControlled,
      isPrescriptionOnly: isPrescriptionOnly ?? this.isPrescriptionOnly,
      size: size ?? this.size,
      color: color ?? this.color,
      season: season ?? this.season,
      supplierId: supplierId ?? this.supplierId,
    );
  }
}
